stores/stores.go

```diff
@@ -521,3 +524,83 @@ func HasSopsTopLevelKey(branch sops.TreeBranch) bool {
 	}
 	return false
 }
+
+// ConvertStructToMap recursively converts a structure to a map[string]interface{} representation while
+// respecting all mapstructure tags on the source structure. This is useful when converting complex structures
+// to a map suitable for use with the Flatten function.
+//
+// Note: this will only emit the public fields of a structure, private fields are ignored entirely.
+func ConvertStructToMap(input interface{}) (map[string]interface{}, error) {
```

Comment:
This function is currently not used.

You need to:
1. modify `FlattenMetadata` in flatten.go to use this function;
2. modify `UnflattenMetadata` in flatten.go to use `mapstructure.WeakDecode`;
3. remove the calls to `DecodeNonStrings`, since they should be no longer needed.

Probably `EncodeNonStrings` should be changed as well to traverse the structure and call `ValueToString` for every non-string value, so that `EncodeNonStrings` no longer needs to know about the exact metadata structure.

Reply:
Oops! Restored ConvertStructToMap and WeakDecode per original patch.  DecodeNonStrings is now removed.

Presently, EncodeNonStrings considers only two cases: mac_only_encrypted and shamir_threshold.  Are you thinking of replacing that with something like:
```golang
	for k, v := range m {
		if _, ok := v.(string); ! ok {
			m[k] = stores.ValueToString(v)
		}
	}
```
?